File: 211/02-2025/Lab05/LinkedQueue.java
// Author: Java Foundations
public class LinkedQueue<T> implements QueueADT<T> {
    private int count;
    private LinearNode<T> head, tail;

    public LinkedQueue() {
        count = 0;
        head = tail = null;
    }

    public void enqueue(T element) {
        LinearNode<T> node = new LinearNode<T>(element);
        if (isEmpty())
            head = node;
        else
            tail.setNext(node);

        tail = node;
        count++;
    }

    public T dequeue() throws EmptyCollectionException {
        if (isEmpty())
            throw new EmptyCollectionException("queue");

        T result = head.getElement();
        head = head.getNext();
        count--;

        if (isEmpty())
            tail = null;

        return result;
    }
   
    public T first() throws EmptyCollectionException {
        if (isEmpty())
        		throw new EmptyCollectionException("Queue is empty");
        return head.getElement();
    }

    public boolean isEmpty() {
        return count==0;
    }
    public int size() {
        return count;
    }

    public String toString() {
        StringBuilder s = new StringBuilder();
        LinearNode<T> current = head;
        while (current != null) {
        		s.append(current.getElement());
        		s.append("\n");
        		current = current.getNext();
        }
        return s.toString();
    }
    
    @Override
	public boolean equals(LinkedQueue<T> list) {
		// Lab exercise 1.  Fill in this method.
		return true;
	}

	public LinkedQueue<T> reverse() {
		// Lab exercise 2.  Fill in this method.
		return null;
	}
}

File: 256/211-Lab05/linked_queue.py
from linear_node import *
from raccoon_exceptions import *

class LinkedQueue(object):
    def __init__(self) -> None:
        self.count: int = 0
        self.head: LinearNode = None
        self.tail: LinearNode = None

    def is_empty(self) -> bool:
        return self.count == 0

    def size(self) -> int:
        return self.count

    def enqueue(self, element: T):
        node: LinearNode = LinearNode(element)
        if self.is_empty():
            self.head = node
        else:
            self.tail.set_next(node)

        self.tail = node
        self.count += 1

    def dequeue(self) -> T:
        if self.is_empty():
            raise EmptyException("Queue is empty. Nothing to dequeue")

        result: T = self.head.get_element()
        self.head = self.head.get_next()
        self.count -= 1

        if self.is_empty():
            self.tail = None
        return result

    def first(self) -> T:
        if self.is_empty():
            raise EmptyException("Queue is empty")

        return self.head.get_element()

    def __str__(self) -> str:
        msg: str = ""
        current: LinearNode = self.head

        while current is not None:
            msg += str(current.get_element()) + "\n"
            current = current.get_next()

        return msg
